FSPT-542: Authentication ADR (#33)
communitiesuk/funding-service-requests-for-comments: adr/2025-07-24-authentication.md

```
Title: Authentication for MVP
Owner: Marc Usher
Collaborator(s): Platform Team
Created on: 2025-08-15
Status: Final
Finalised on: 2025-08-15
```

## Overview

> The new Funding Service platform is conceptually split into two services — Deliver Grant Funding, for internal users, and Access Grant Funding, for external and end users. 

Reviewer: I think we might want to refer to user profiles rather than generic terms like 'internal users' and 'end users'.

```suggestion
The new Funding Service platform is conceptually split into two services — Deliver Grant Funding, for grant team members and funding service admin users, and Access Grant Funding, for grant applicants and recipients. 
```

Author: Yeah good shout thank you, will update!

The service will have a number of users, including Funding Service admins, grant teams, other government department (OGD) teams, assessors, grant applicants, grant recipients and Section 151 Officers. Our authentication methods need to be secure, robust and reliable, meeting service standards, secure by design principles and MHCLG Cyber Security requirements, while also meeting the user needs for these internal and external user groups.

This ADR touches on briefly how, in the current service, authentication is closely tied with authorisation, but you can [read the recent authorisation ADR](https://github.com/communitiesuk/funding-service-requests-for-comments/blob/main/adr/2025-05028-authorisation-model.md) for more on Deliver/Access Grant Funding's proposed authorisation model.

## What is the current state?

> Authentication on our current services (Apply, Assess, Find, Submit) is currently a mix of two authentication methods, magic link and federated authentication via Microsoft Azure AD using MHCLG's Azure AD tenant. All authentication is handled by what was previously a separate Funding Service-built microservice, 'Authenticator', which is now part of the consolidated 'Pre-award' codebase.

Reviewer: ```suggestion
Authentication on our current services (Apply, Assess, Find, Submit) is currently a mix of two authentication methods: magic links, and federated authentication via Microsoft Azure AD using MHCLG's Azure AD tenant. All authentication is handled by what was previously a separate Funding Service-built microservice, 'Authenticator', which is now part of the consolidated 'Pre-award' codebase.
```

### Apply

> Applicants are sent direct links to grant application pages and on visiting those pages are prompted input their email to generate 'magic links'. They receive an email with a link which allows them to authenticate and access a session to start or complete the application process. Due to the permissions model for Apply, all applications are therefore tied to just one user and cannot be shared or collaborated on by users in the same organisation.

Reviewer: ```suggestion
Applicants are sent direct links to grant application pages and on visiting those pages are prompted to input their email to generate 'magic links'. They receive an email with a link which allows them to authenticate and access a session to start or complete the application process. Due to the permissions model for Apply, all applications are therefore tied to just one user and cannot be shared or collaborated on by users in the same organisation.
```

Based on their email domain we map them to a Local Authority and determine who they are able to submit returns for. There are a number of users who need to be able to submit for multiple organisations, however, which requires manually adding them to a JSON secret in AWS Parameter Store with a mapping of their email address and the organisations for whom they will be submitting monitoring reports.

### Delta

While not a Funding Service built service, Delta's authentication methods are useful to consider. Delta uses the OAuth 2 Authorization Code flow which allows them to use a standard OAuth client library in Delta. Unauthenticated users are redirected to its own auth service and shown a login page where they can provide a username and password as a standard form submission, or with Single Sign On (SSO) through Azure AD.

For more on Delta's authentication method, [read their README guidance](https://github.com/communitiesuk/delta-auth-service/tree/main/auth-service#oauth-flow).

## What are the problems with the current state?

> Authentication for the current services involves a number of long, error-prone manual processes and is a recurring pain point for users, frequently coming up as issue in service desk support tickets ([see examples on Confluence](https://mhclgdigital.atlassian.net/wiki/spaces/FS/pages/569344026/Qualitative+Support+ticket+analysis+for+February+28th+-+May+30th+2025#6.4-Authentication-log-in)).

Reviewer: ```suggestion
Authentication for the current services involves a number of long, error-prone manual processes and is a recurring pain point for users, frequently coming up as an issue in service desk support tickets ([see examples on Confluence](https://mhclgdigital.atlassian.net/wiki/spaces/FS/pages/569344026/Qualitative+Support+ticket+analysis+for+February+28th+-+May+30th+2025#6.4-Authentication-log-in)).
```

- All users (including guest users) in our Azure AD are required to use Multi-Factor Authentication (MFA) when authenticating. For users of Submit, changing their MFA device or losing it often results in them not being able to authenticate, and a request has to be made to our MHCLG IT team to reset it, along with verification that the request is legitimate. All of this takes time and is a process outside of Funding Service's control.
- There is also regular confusion about the link to the Submit service, with users authenticating via Azure AD and expecting it to be present in the Microsoft 'Apps' window rather than then going directly to the service URL.
- Use of the federated Azure AD authentication and light-touch email domain mapping for permissions is also brittle on the Submit service, with new users needing to be added to the relevant Azure AD Security Groups for each monitoring round, and users who need to submit returns for multiple local authorities needing to be manually added to a secrets JSON in AWS which grants additional permissions.

> - In Assess, the need to add user emails to specific Azure AD Security Groups for each new grant (and the setting up of these security groups in the first place) is a long and manual process. Tying permissions to these groups also leads to issues on non-production environments when needing to test grants at different permission levels in end to end tests using the same email addresses leads to issue with the basic authentication sidecar protecting these environments, as the sheer number of roles returned causes issues (see the [Jira ticket documenting this issue](https://mhclgdigital.atlassian.net/browse/FLS-1433)).

Reviewer: Nice representation of some of the core/common issues 👍

## What authentication methods are we implementing for MVP?

> We are not proposing radically different authentication methods for Access or Deliver Grant Funding, but rather looking to solve or avoid the issue listed above.

Reviewer: ```suggestion
We are not proposing radically different authentication methods for Access or Deliver Grant Funding, but rather looking to solve or avoid the issues listed above.
```

### Deliver Grant Funding (internal users)

We will use federated authentication via MHCLG's Azure AD tenant to control access to Deliver Grant Funding. This part of the service is the administration interface for internal users, all of whom will have existing @communities.gov.uk email addresses or, in the case of future Assessors, be familiar with authenticating via Microsoft and will likely already be guest users in Azure AD.

With one exception, we are not proposing using Azure AD Security Groups to sync or manage permission groups, as this very manual process has proven to be a pain point for Funding Service when onboarding grants or setting up monitoring reports. Instead, Role Based Access Control is controlled by the database in the application ([read the recent authorisation ADR](https://github.com/communitiesuk/funding-service-requests-for-comments/blob/main/adr/2025-05028-authorisation-model.md) for more on how we propose to manage permissions within Access and Deliver Grant Funding). 

> The one exception is the pre-existing `FSD_ADMIN` Security Group, whose members include most of the Funding Service. This group will be used to identify SSO users who should have Funding Service Admin permissions within the platform when they sign in, and will be kept as the single source of truth.

Reviewer: I have a feeling in the back of my head that we may need to disambiguate this further at some point - we may need funding service form designers and funding service platform admins (developers) as separate permission levels.

Author: Yeah this is a good point, it might be that we end up needing two (small) security groups to manage this going forward, or that FS form designers are designated in the platform by the FS admins rather than coming from a security group. I can add an extra line or two to highlight we'd maybe need to deal with this in the future.
